modules/xfem/src/userobjects/MeshCut3DUserObject.C

```diff
@@ -0,0 +1,1106 @@
+      unsigned int counter = -1;
+      for (; elem_it2 != elem_end; ++elem_it2)
+      {
+        ++counter;
+
+        if (counter != special_boundary_elem_id[i])
+        {
+          Elem * cut_elem2 = *elem_it2;
+          for (unsigned int k = 0; k < 3; ++k)
+          {
+            unsigned int node3 = cut_elem2->get_node(k)->id();
+            unsigned int node4 = cut_elem2->get_node((k + 1 <= 2) ? k + 1 : 0)->id();
+            if (node3 > node4)
+              std::swap(node3, node4);
+
+            if (node1 == node3 && node2 == node4)
+            {
+              is_edge_inside = 1;
+              goto endloop;
+            }
+          }
+        }
+      }
+    endloop:
+      if (is_edge_inside == 0)
+      {
+        // store good edges
+        _boundary_edges.push_back(new CutEdge);
+        _boundary_edges[_boundary_edges.size() - 1]->id1 = node1;
+        _boundary_edges[_boundary_edges.size() - 1]->id2 = node2;
+      }
+      else
+      {
+        // this is a bad edge; remove it from existing edge list
+        for (unsigned int k = 0; k < _boundary_edges.size(); ++k)
+        {
+          if (_boundary_edges[k]->id1 == node1 && _boundary_edges[k]->id2 == node2)
+          {
+            _boundary_edges.erase(_boundary_edges.begin() + k);
+            break;
+          }
+        }
+      }
+    }
+  }
+}
+
+void
+MeshCut3DUserObject::sortBoundaryNodes()
+{
+  _boundary.clear();
+
+  for (unsigned int i = 0; i < _boundary_edges.size(); ++i)
+  {
+    unsigned int node1 = _boundary_edges[i]->id1;
+    unsigned int node2 = _boundary_edges[i]->id2;
+    _boundary_map.at(node1)->neighbor.push_back(node2);
+    _boundary_map.at(node2)->neighbor.push_back(node1);
+  }
+
+  std::map<int, CutPoint *>::iterator it = _boundary_map.begin();
+  while (it != _boundary_map.end())
+  {
+    if (it->second->neighbor.size() != 2)
+      mooseError("Boundary points in cutter mesh must have exactly two neighbors");
+    ++it;
+  }
+
+  unsigned int node1 = _boundary_edges[0]->id1;
+  unsigned int node2 = _boundary_edges[0]->id2;
+  _boundary.push_back(node1);
+  _boundary.push_back(node2);
+  for (unsigned int i = 0; i < _boundary_edges.size() - 1; ++i)
+  {
+    unsigned int node3 = _boundary_map.at(node2)->neighbor[0];
+    unsigned int node4 = _boundary_map.at(node2)->neighbor[1];
+    if (node3 == node1)
+    {
+      _boundary.push_back(node4);
+      node1 = node2;
+      node2 = node4;
+    }
+    else if (node4 == node1)
+    {
+      _boundary.push_back(node3);
+      node1 = node2;
+      node2 = node3;
+    }
+    else
+      mooseError("Discontinuity in cutter boundary");
+  }
+}
+
+Real
+MeshCut3DUserObject::findDistance(unsigned int node1, unsigned int node2)
+{
+  Node * n1 = _cut_mesh->node_ptr(node1);
+  Point & p1 = *n1;
+  Node * n2 = _cut_mesh->node_ptr(node2);
+  Point & p2 = *n2;
+  Real distance = std::sqrt(std::pow(p1(0) - p2(0), 2) + std::pow(p1(1) - p2(1), 2) +
```

Comment:
`Points` already calculate distances between themselves. Don't reinvent the wheel.